---
title: "looking for IL11"
author: "Miguel Ramon"
date: "`r Sys.Date()`"
citations:  
output:
  html_document: default
  pdf_document: default
---

```{r setup}

knitr::opts_knit$set(root.dir = '/home/miguel/IBE-UPF/PhD/NEOPLASY_PRIMATES/')
knitr::opts_chunk$set(engine.path = list(
python = '/home/miguel/IBE-UPF/PhD/NEOPLASY_PRIMATES/TreeCluster/venv/bin/python'))
getwd()

```

## Directory and functiondefinition
```{r dirdef}

workingDir <- "/home/miguel/IBE-UPF/PhD/NEOPLASY_PRIMATES/"
dataDir <- file.path(workingDir, "Data")
resultsDir <- file.path(workingDir,"Out") 

getwd()

createDir <- function(directory) {
  if (!file.exists(directory)) {
    dir.create(directory, recursive = TRUE)
  }
}

robust_scale <- function(x) {
  (x - median(x)) / IQR(x)
}

```

### Seed and library loading

```{r,warning=FALSE,message=FALSE}

set.seed(1998)

# General use libraries
library(readr)
library(phytools)
library(dplyr)
library(castor)
library(tidyr)
library(dplyr)
library(purrr)
library(data.table)


library(ggtree)
library(ggplot2)
library(ggrepel)
library(ggExtra)

```
# Load bootstrap objects
```{r boot-objects}

# File paths
caas_dir <- file.path(resultsDir, "2.CAAS/Ranked_def")
caas_path <- file.path(caas_dir, "1.Discovery")
resamp_path <- file.path(caas_dir, "2.Resample/ranked")
boot_path <- file.path(caas_dir,"3.Bootstrap/1.Preproc-setup-dfs/ranked")
random.resamp_path <- file.path(caas_dir,"2.Resample/random")
random.boot_path <- file.path(caas_dir,"3.Bootstrap/1.Preproc-setup-dfs/random")

# KEEPING ONLY SIGNIFICANT VALUES
## Read caas files
caas_files <- list.files(caas_path, pattern = "\\.tab$", full.names = TRUE)
caas_dfs <- lapply(caas_files, read_tsv)

## Keep significant
caas_dfs <- lapply(caas_dfs, function(df) {
  df <- df %>% filter(as.numeric(Pvalue) <= 0.05)
  df$GenePos <- paste(df$Gene, df$Position, sep = "@")
  return(df)
})

# Basename caas_files up to the hyphen
caas_files <- basename(caas_files)
caas_files <- sapply(caas_files, function(x) strsplit(x, "-")[[1]][1])

## Rename elements within the list
names(caas_dfs) <- caas_files

# Read resample files
resamp_files <- list.files(resamp_path, pattern = "\\.resamp$", full.names = TRUE)
resamp_dfs <- lapply(resamp_files, read_tsv, col_names = FALSE)
## Change column names to something more informative
resamp_dfs <- lapply(resamp_dfs, function(df) {
  colnames(df) <- c("Permutation", "ForegroundSpecies", "BackgroundSpecies")
  return(df)
})

## Rename elements within the list
names(resamp_dfs) <- caas_files

# Random resample dfs
random_resamp_files <- list.files(random.resamp_path, pattern = "\\.resamp$", full.names = TRUE)
random_resamp_dfs <- lapply(random_resamp_files, read_tsv, col_names = FALSE)
random_resamp_dfs <- lapply(random_resamp_dfs, function(df) {
  colnames(df) <- c("Permutation", "ForegroundSpecies", "BackgroundSpecies")
  return(df)
})

names(random_resamp_dfs) <- caas_files

# Read bootstrap files
boot_files <- list.files(boot_path, pattern = "\\.tab$", full.names = TRUE)
boot_dfs <- lapply(boot_files, read_tsv)

boot_dfs <- lapply(boot_dfs, function(df) {
  colnames(df) <- c("GenePos", "Counts", "Cycles", "adj.pval", "Bootstrap_greater", "Trait")
  return(df)
})

## Change column names to something more informative
boot_dfs <- lapply(boot_dfs, function(df) {
  df <- df %>%
    filter(as.numeric(adj.pval) <= 0.05) 
  return(df)
})

names(boot_dfs) <- caas_files

# Random boot dfs
random_boot_files <- list.files(random.boot_path, pattern = "\\.tab$", full.names = TRUE)
random_boot_dfs <- lapply(random_boot_files, read_tsv)

random_boot_dfs <- lapply(random_boot_dfs, function(df) {
  colnames(df) <- c("GenePos", "Counts", "Cycles", "adj.pval", "Bootstrap_greater", "Trait")
  return(df)
})

# Only significant
random_boot_dfs <- lapply(random_boot_dfs, function(df) {
  df <- df %>%
    filter(as.numeric(adj.pval) <= 0.05) 
  return(df)
})

names(random_boot_dfs) <- caas_files

# Join the dataframes
setup_df <- lapply(names(caas_dfs), function(name) {
  caas_df <- caas_dfs[[name]]
  boot_df <- boot_dfs[[name]]
  merged_df <- merge(caas_df, boot_df, by = "GenePos", all.x = FALSE)
  return(merged_df)
})

names(setup_df) <- caas_files

# Random setup df
random_setup_df <- lapply(names(caas_dfs), function(name) {
  caas_df <- caas_dfs[[name]]
  boot_df <- random_boot_dfs[[name]]
  merged_df <- merge(caas_df, boot_df, by = "GenePos", all.x = FALSE)
  return(merged_df)
})

# Load config data as b0
config_path <- file.path(caas_dir, "4.Traitfiles")
config_files <- list.files(config_path, pattern = "\\.tab$", full.names = TRUE)
config_dfs <- lapply(config_files, read_tsv, col_names = FALSE)
names(config_dfs) <- caas_files

## New resample dfs
resamp_config_dfs <- lapply(config_dfs, function(df) {
  df <- df %>% 
    # Group by second column
    group_by(X2) %>%
    # Collapse first column into a single string
    summarise(X1 = paste(X1, collapse = ","), .groups = 'drop') %>%
    # In column X1, change 0 to BackgroundSpecies and 1 to ForegroundSpecies
    mutate(X2 = ifelse(X2 == "0", "BackgroundSpecies", "ForegroundSpecies")) %>%
    # Pivot the columns
    pivot_wider(names_from = X2, values_from = X1) %>%
    # Create column Permutation with values b_0
    mutate(Permutation = "b_0") %>%
    # Reorder columns
    select(Permutation, ForegroundSpecies, BackgroundSpecies)
  return(df)
})

# Bind resample dfs with resamp_config_dfs as the first element
resamp_dfs <- lapply(names(resamp_dfs), function(name) {
  resamp_df <- resamp_dfs[[name]]
  resamp_config_df <- resamp_config_dfs[[name]]
  merged_df <- bind_rows(resamp_config_df, resamp_df)
  return(merged_df)
})

names(resamp_dfs) <- names(resamp_config_dfs)

# Do the same for the random resample dfs
random.resamp_dfs <- lapply(names(random_resamp_dfs), function(name) {
  resamp_df <- random_resamp_dfs[[name]]
  resamp_config_df <- resamp_config_dfs[[name]]
  merged_df <- bind_rows(resamp_config_df, resamp_df)
  return(merged_df)
})

names(random.resamp_dfs) <- names(resamp_config_dfs)

# Load permutation gene maps
permutation_gene_maps <- lapply(names(caas_dfs), function(name) {
  file_path <- file.path(caas_dir, "3.Bootstrap/3.Permutation_gene_maps", paste0(name, ".tsv"))
  df <- read_tsv(file_path)
  return(df)
})

names(permutation_gene_maps) <- names(caas_dfs)

# Filter permutation gene maps to only include IL11
permutation_gene_maps <- lapply(permutation_gene_maps, function(df) {
  df <- df %>%
    filter(Gene == "IL11")
  return(df)
})

# Load random permutation gene maps
random_permutation_gene_maps <- lapply(names(caas_dfs), function(name) {
  file_path <- file.path(caas_dir, "3.Bootstrap/8.Random_permutation_gene_maps")
  files <- list.files(file_path, pattern = "tsv", full.names = TRUE)
  df <- read_tsv(files)
  return(df)
})

names(random_permutation_gene_maps) <- names(caas_dfs)

# Gene length information
gene_lengths <- read_tsv(file.path(dataDir, "2.Alignments/gene_lengths.tab"), col_names = c("Gene", "Length"))
gene_lengths <- gene_lengths %>%
  # Keep only the gene name (until the first dot)
  mutate(Gene = sub("\\..*", "", Gene))

# Filter random permutation gene maps to only include IL11
random_permutation_gene_maps <- lapply(random_permutation_gene_maps, function(df) {
  df <- df %>%
    group_by(Gene, Permutation, ForegroundSpecies, BackgroundSpecies) %>%
    summarise(TotalCAAS = n(), .groups = 'drop') %>%
    # Join with gene_lengths to get the gene length
    left_join(gene_lengths, by = "Gene") %>%
    # Correct the number of CAAS by the gene length
    mutate(TotalCAAS_corr = TotalCAAS / Length) %>%
    # Remove NA values
    filter(!is.na(Gene)) %>%
    filter(Gene == "IL11")
  return(df)
})

```

# Add cancer stuff
```{r permulation_table_extract}
# Load cancer data
cancer_data <- read_csv(file.path(dataDir, "1.Cancer_data/Neoplasia_species360/clean_primate_traits.csv"))

cancer_data_filt <- cancer_data %>%
  dplyr::select(species, neoplasia_prevalence) %>%
  dplyr::rename(Species = species, NeoplasiaPrevalence = neoplasia_prevalence) %>%
  mutate(CancerType = case_when(
    NeoplasiaPrevalence <= quantile(NeoplasiaPrevalence, 0.25, na.rm=TRUE) ~ "Low",
    NeoplasiaPrevalence >= quantile(NeoplasiaPrevalence, 0.75, na.rm=TRUE) ~ "High",
    TRUE ~ "Intermediate"))

# See how much cancer permutated species have
cancer_apport_f <- function(df) {
cancer_apport_df <- df %>%
  select(Gene, Permutation, ForegroundSpecies, BackgroundSpecies, TotalCAAS, Length) %>%
  # Remove duplicate Bootstrap_greater
  distinct() %>%
  mutate(ForegroundSpecies = strsplit(ForegroundSpecies, ",")) %>%
  mutate(BackgroundSpecies = strsplit(BackgroundSpecies, ",")) %>%
  unnest(cols = c(ForegroundSpecies, BackgroundSpecies)) %>%
  # Melt the dataframe to long format
  reshape2::melt(id.vars =  c("Gene", "Permutation", "TotalCAAS", "Length"), measure.vars = c("ForegroundSpecies", "BackgroundSpecies")) %>%
  dplyr::rename(Species = value, ConfigType = variable) %>%
  drop_na() %>%
  # Left join with cancer_data_filt
  left_join(cancer_data_filt, by = c("Species" = "Species")) %>%
  # For ConfigType, change ForegroundSpecies to 1 and BackgroundSpecies to 0
  mutate(ConfigType = ifelse(ConfigType == "ForegroundSpecies", 1, 0)) %>%
  # For NeoplasiaPrevalence, change HighNP to 1, Normal to 2, and LowNP to 0
  mutate(CancerType = case_when(
    CancerType == "High" ~ 1,
    CancerType == "Intermediate" ~ 2,
    CancerType == "Low" ~ 0
  )) %>%
  # If ConfigType is 1 and CancerType is 1, then IsCancer is IsFG, if ConfigType is 0 and CancerType is 0, then IsCancer is IsBG, else IsCancer is FALSE
  mutate(IsCancer = ifelse(ConfigType == 1 & CancerType == 1, "IsFG", ifelse(ConfigType == 0 & CancerType == 0, "IsBG", "FALSE"))) %>%
  mutate(IsWrong = ifelse(ConfigType == 1 & CancerType == 0, "IsFG", ifelse(ConfigType == 0 & CancerType == 1, "IsBG", "FALSE"))) %>%
  dplyr::select(Gene, Permutation, Species, IsCancer, IsWrong, TotalCAAS, Length) %>%
  # # Group by Permutation
  group_by(Gene,Permutation) %>%
  # # Per group, count the situation where IsCancer is IsFG, IsBG, and FALSE
  summarise(NPTop = sum(IsCancer == "IsFG"),
            NPBottom = sum(IsCancer == "IsBG"),
            NPFALSE = sum(IsCancer == "FALSE"),
            NPWrongTop = sum(IsWrong == "IsFG"),
            NPWrongBottom = sum(IsWrong == "IsBG"),
            # NPWrong is the sum of the wrong top and wrong bottom
            NPWrong = NPWrongTop + NPWrongBottom,
            # Overall bias is the sum of the wrong and false
            NPOverallBias = NPWrong + NPFALSE,
            .groups = 'drop')
}

# Apply cancer_apport_f to all permutation_gene_maps using lapply
cancer_apport_dfs <- lapply(permutation_gene_maps, cancer_apport_f)
random_cancer_apport_dfs <- lapply(random_permutation_gene_maps, cancer_apport_f)

```

# All other traits

```{r}
# Load sp2fam
sp2fam <- read_tsv(file.path(dataDir, "1.Cancer_data/Neoplasia_species360/sp2fam_wo_groups.tab"))

# Extract unique species present in resamp_dfs
unique_species <- lapply(resamp_dfs, function(df) {
  df %>%
    select(ForegroundSpecies) %>%
    # Split ForegroundSpecies by comma
    mutate(ForegroundSpecies = strsplit(ForegroundSpecies, ",")) %>%
    # Unnest ForegroundSpecies
    unnest(cols = c(ForegroundSpecies)) %>%
    # In both ForegroundSpecies, remove duplicates
    distinct() %>%
    # Extract the list
    pull(ForegroundSpecies)
})

unique_species <- unique(unlist(unique_species))

# Load all other traits
all_traits <- read_tsv(file.path(dataDir, "6.Other-primate-phenotypes/Supplementary_Tables_1_-_Non-human_primates_phenomic_dataset 2.tsv"))

# Left join family information
all_traits <- all_traits %>%
  left_join(sp2fam, by = c("SpeciesBROAD" = "Species")) %>%
  # Remove species not in unique_species
  filter(SpeciesBROAD %in% unique_species) %>%
  # Set AOB, Epithalamus, Subthalamic.Nucleus and XII as numeric
  mutate(across(c("AOB", "Epithalamus", "Subthalamic.Nucleus", "XII"), as.numeric))

# Separate numeric and character columns in two different dataframes
all_traits_numeric <- all_traits %>%
  # Select only numeric columns while keeping SpeciesBROAD and GroupName
  select(SpeciesBROAD, Family, where(is.numeric)) %>%
  dplyr::rename(Species = SpeciesBROAD)
# 
# all_traits_character <- all_traits %>%
#   # Select only numeric columns while keeping SpeciesBROAD and GroupName
#   select(SpeciesBROAD, Family, where(is.character)) %>%
#   select(-GroupName) %>%
#   dplyr::rename(Species = SpeciesBROAD)

# Tag as Low, Intermediate, High every trait in the numeric dataframe
all_traits_numeric_tagged <- all_traits_numeric %>%
  mutate(across(where(is.numeric), ~case_when(
    . <= quantile(., 0.25, na.rm=TRUE) ~ "0",
    . >= quantile(., 0.75, na.rm=TRUE) ~ "1",
    # If NA, unknown
    is.na(.) ~ NA,
    TRUE ~ "2"
  )))

# Remove columns where all values are NA
all_traits_numeric_tagged <- all_traits_numeric_tagged %>%
  select(where(~!all(is.na(.))))
```


```{r map_to_traits}

trait_apport_f <- function(df) {
  trait_apport_df <- df %>%
    select(Gene, Permutation, ForegroundSpecies, BackgroundSpecies, TotalCAAS, Length) %>%
    # Remove duplicate Bootstrap_greater
    distinct() %>%
    mutate(ForegroundSpecies = strsplit(ForegroundSpecies, ",")) %>%
    mutate(BackgroundSpecies = strsplit(BackgroundSpecies, ",")) %>%
    unnest(cols = c(ForegroundSpecies, BackgroundSpecies)) %>%
    # Melt the dataframe to long format
    reshape2::melt(id.vars =  c("Gene", "Permutation", "TotalCAAS", "Length"), measure.vars = c("ForegroundSpecies", "BackgroundSpecies")) %>%
    dplyr::rename(Species = value, ConfigType = variable) %>%
    drop_na() %>%
    # Left join with cancer_data_filt
    left_join(all_traits_numeric_tagged, by = c("Species" = "Species")) %>%
    # For ConfigType, change ForegroundSpecies to 1 and BackgroundSpecies to 0
    mutate(ConfigType = ifelse(ConfigType == "ForegroundSpecies", 1, 0))
  
  # For every trait, check if Trait is 1 and ConfigType is 1, then IsTrait is IsFG, if ConfigType is 0 and Trait is 0, then IsTrait is IsBG, else IsTrait is FALSE
  for (trait in colnames(trait_apport_df)[8:ncol(trait_apport_df)]) {
    trait_name <- paste0("Is", trait)
    trait_apport_df <- trait_apport_df %>%
      # Do this with something other than mutate
      # Preferibly, replace directly the columns that show the phenos
      mutate(trait = ifelse(ConfigType == 1 & .data[[trait]] == "1", "IsFG", ifelse(ConfigType == 0 & .data[[trait]] == "0", "IsBG", "FALSE"))
      )
    trait_apport_df <- trait_apport_df %>%
      dplyr::select(Gene, Permutation, Species, IsTrait, TotalCAAS, Length) %>%
      # # Group by Permutation
      group_by(Gene,Permutation) %>%
      # # Per group, count the situation where IsTrait is IsFG, IsBG, and FALSE
      summarise(NPTop = sum(IsTrait == "IsFG"),
                NPBottom = sum(IsTrait == "IsBG"),
                NPFALSE = sum(IsTrait == "FALSE"),
                .groups = 'drop')
  }
  # Run the trait_apport function
  return(trait_apport_df)
}

trait_apport_dfs <- lapply(permutation_gene_maps, trait_apport_f)

```
```{r load_mammals}

# Load mammal data
mammal_data <- read_tsv(file.path(dataDir, "2.Alignments/Mammals_alignments/mammal_genes_template.txt"))

mammal_data_prunned <- mammal_data %>%
  select(Species, `Common name`, `Assembly name`) %>%
  dplyr::rename(Species = Species, CommonName = `Common name`, AssemblyName = `Assembly name`) %>%
  # Add vs_ to the Assembly name
  mutate(AssemblyName = paste0("vs_", AssemblyName)) %>%
  # Binarize the Species usign a _
  mutate(Species = sub(" ", "_", Species)) %>%
  # Do the same with common name
  mutate(CommonName = sub(" ", "_", CommonName))

# Export the table
write_tsv(mammal_data_prunned, file.path(dataDir, "2.Alignments/Mammals_alignments/mammal_genes_template_prunned.txt"))

name2assembly <- mammal_data_prunned %>%
  select(Species, AssemblyName) %>%
  # Remove duplicates
  distinct()

# Export the conversion table
write_tsv(name2assembly, file.path(dataDir, "2.Alignments/Mammals_alignments/name2assembly.txt"), col_names = FALSE)

# Extract both columns as separate files
write.table(name2assembly$Species, file.path(dataDir, "2.Alignments/Mammals_alignments/mammals_species.txt"), col.names = FALSE, row.names = FALSE, quote = FALSE)
write.table(name2assembly$AssemblyName, file.path(dataDir, "2.Alignments/Mammals_alignments/mammals_assembly.txt"), col.names = FALSE, row.names = FALSE, quote = FALSE)

# Load primates alignments
primate_alignments <- list.files(file.path(dataDir, "2.Alignments/Primate_alignments"), pattern = "phy")
## To dataframe
primate_alignments_df <- data.frame(Primate_Alignment = primate_alignments)
primate_alignments_df <- primate_alignments_df %>%
  # Keep only the gene name (until the first dot)
  mutate(Primate_Alignment = sub("\\..*", "", Primate_Alignment))

# Load mammals alignments
mammal_alignments <- list.files(file.path(dataDir, "2.Alignments/Mammals_alignments/MultipleCodonAlignments"), pattern = "fasta.gz")
# To dataframe
mammal_alignments_df <- data.frame(Mammal_Alignment = mammal_alignments)
mammal_alignments_df <- mammal_alignments_df %>%
  # Keep only the gene name (separate by dots and keep second element)
  mutate(Mammal_Alignment = sapply(strsplit(Mammal_Alignment, "\\."), function(x) x[2]))

# Compare the two dataframes by extracting the gene names not present in the primate alignments
compared_alignments <- anti_join(mammal_alignments_df, primate_alignments_df, by = c("Mammal_Alignment" = "Primate_Alignment"))
## Same the other way around
compared_alignments2 <- anti_join(primate_alignments_df, mammal_alignments_df, by = c("Primate_Alignment" = "Mammal_Alignment"))

```

