---
title: "Ensembl annotation extraction"
author: "Miguel Ramon"
date: "`r Sys.Date()`"
citations:   
output:
  html_document: default
  pdf_document: default
---

## 1. Setup & Directory Definition

This section sets the working directory, loads common functions, and defines data and results folders.

```{r setup}

knitr::opts_knit$set(root.dir = "/home/miguel/IBE-UPF/PhD/NEOPLASY_PRIMATES")
getwd()

```

```{r dirdef, message=FALSE, warning=FALSE}

workingDir <- getwd()
dataDir <- file.path(workingDir, "Data")
resultsDir <- file.path(workingDir,"Malignancy_Primates/Out/")

# Load common objects, functions and libraries
source(file.path(workingDir, "Malignancy_Primates/Scripts/functions/commons.R"))
source(file.path(workingDir, "Malignancy_Primates/Scripts/functions/primate_functions.R"))
source(file.path(workingDir, "Malignancy_Primates/Scripts/functions/primate_objects.R"))
source(file.path(workingDir, "Malignancy_Primates/Scripts/functions/dunn_modified.R"))

```

# Ensembl gene data

```{r ensembl}

gene_info_files <- list.files(path = file.path(dataDir, "2.Alignments"), pattern = "gene_info_", full.names = TRUE, recursive = FALSE)
gene_info <- lapply(gene_info_files, read_tsv, col_names = FALSE)
gene_lengths.df <- read_tsv(file.path(dataDir, "2.Alignments/gene_lengths.tab"), col_names = FALSE)
gene_lengths <- gene_lengths.df %>%
  dplyr::rename(Gene = X1, Length = X2) %>%
  # Keep string in "Gene" up to the first dot
  mutate(Gene = str_extract(Gene, "^[^.]+"))

gene_info_data <- gene_info[[1]] %>%
  dplyr::filter(X3 == "mRNA") %>%
  dplyr::select(X1, X4, X5, X7, X9) %>%
  dplyr::rename(chr = X1, start = X4, end = X5, strand = X7, gene = X9) %>%
  arrange(chr, start) %>%
  # Modify gene to keep string between = and ;
  mutate(gene = str_extract(gene, "(?<=\\=)[^;]+")) %>%
  # Left_join gene_info[[2]] by gene=X2
  left_join(gene_info[[2]], by = c("gene" = "X2")) %>%
  dplyr::select(-gene) %>%
  dplyr::rename(gene = X1) %>%
  # Remove NA
  filter(!is.na(gene))

# Check which genes from gene_list are not in gene_info_data
non_coincident <- gene_info_data$gene[!gene_lengths$Gene %in% gene_info_data$gene]

# Check repeats in gene_info_data
gene_info_data$gene[duplicated(gene_info_data$gene)]

# Load ensembl data from downloaded mart file. Biomart R function is way too inconsistent.
ens_db <- read.csv(file.path(dataDir, "2.Alignments/ensembl_mart_export.csv"))

gene_coords_g <- ens_db %>%
  dplyr::rename(gene = Gene.name, chr = Chromosome.scaffold.name, start = Gene.start..bp., end = Gene.end..bp., strand = Strand) %>%
  arrange(chr, start) %>%
  # Remove duplicates keeping those were chr is numeric OR X OR Y
  filter(grepl("^[0-9XYMT]+$", chr)) %>%
  # Add chr prefix to chromosome
  mutate(chr = paste0("chr", chr))

# Join to gene_info_data
gene_info_data2 <- gene_info_data %>%
  rbind(gene_coords_g, by = c("gene" = "gene")) %>%
  # Remove rows were strand is "gene"
  filter(strand != "gene") %>%
  # Mutate strand -1 to - and strand 1 to +. Other values should stay the same.
  mutate(strand = ifelse(strand == 1, "+", ifelse(strand == -1, "-", strand))) %>%
  # Natural sort by chromosome then start position
  arrange(
    factor(chr, levels = unique(chr[str_order(chr, numeric = TRUE)])),
    factor(start, levels = unique(start[str_order(start, numeric = TRUE)])),
  )

# Merge, clean, and aggregate gene data
gene_info_data3 <- gene_info_data2 %>%
  left_join(gene_lengths, by = c("gene" = "Gene")) %>%
  dplyr::select(gene, chr, start, end, strand, Length) %>%
  dplyr::rename(length = Length) %>%
  drop_na(length) %>%
  # Convert start, end, and length to numeric
  mutate(start = as.numeric(start),
         end = as.numeric(end),
         length = as.numeric(length)) %>%
  # Group by gene and chromosome and calculate median start and end values (rounded up)
  group_by(gene, chr) %>%
  mutate(start = ceiling(median(start, na.rm = TRUE)),
         end = ceiling(median(end, na.rm = TRUE))) %>%
  # Keep only one row per gene/chromosome group
  filter(row_number() == 1) %>%
  ungroup() %>%
  # Natural sort by chromosome then by start position
  arrange(
    factor(chr, levels = unique(chr[stringr::str_order(chr, numeric = TRUE)])),
    start
  )

# Identify genes present on both chrX and chrY (pseudoautosomal regions)
pseudo_genes <- gene_info_data3 %>%
  group_by(gene) %>%
  filter(any(chr == "chrX") & any(chr == "chrY")) %>%
  ungroup() %>%
  distinct(gene)

# For genes in the pseudoautosomal regions, keep only the copy in chrX
gene_info_data3_final <- gene_info_data3 %>%
  filter(!(gene %in% pseudo_genes$gene & chr == "chrY"))

# Check non-coincident genes between gene_lengths and gene_info_data3
non_coincident2 <- gene_info_data3_final$gene[!gene_lengths$Gene %in% gene_info_data3_final$gene]
# Add chr data from gene_info_data3 to non_coincident2
non_coincident2_chr <- gene_info_data3_final$chr[gene_info_data3_final$gene %in% non_coincident2]
names(non_coincident2) <- non_coincident2_chr

# > non_coincident2
#      chr3      chr5      chr5      chr7     chr11     chr13      chrX 
#    "BCHE"     "ZFR"   "CCNJL"   "TNPO3"    "THY1"      "F7" "MAGEB10" 

# Check duplicates in gene_info_data3 (NONE)
gene_info_data3_final$gene[duplicated(gene_info_data3_final$gene)]
gene_info_data3_final$chr[duplicated(gene_info_data3_final$gene)]

# Extract the list
path <- file.path(paste0(dataDir,"/2.Alignments")) # add /pattern4/ for pattern4
createDir(path)
write.table(gene_info_data3, file.path(path, "ensembl_genes.output"), quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
  
  


```